File: src-frontend/src/hooks/settings/use_persistent_settings.rs
use leptos::*;
use crate::types::settings::AppSettings;
use crate::services::settings_service::{get_app_settings, update_app_settings, default_settings};
use leptos::spawn_local;

/// Hook for persisting app settings to backend database
/// Returns: (settings, is_hydrated, update_settings)
pub fn use_persistent_settings() -> (
    RwSignal<AppSettings>,
    ReadSignal<bool>,
    Callback<AppSettings>,
) {
    let settings = create_rw_signal(default_settings());
    let (is_hydrated, set_is_hydrated) = create_signal(false);
    
    // Load settings on mount
    create_effect(move |_| {
        if !is_hydrated.get() {
            spawn_local(async move {
                match get_app_settings().await {
                    Ok(loaded_settings) => {
                        settings.set(loaded_settings);
                        set_is_hydrated.set(true);
                    }
                    Err(e) => {
                        web_sys::console::warn_1(&format!("[SettingsPersistence]: failed to load settings from backend, using defaults: {}", e).into());
                        set_is_hydrated.set(true);
                    }
                }
            });
        }
    });
    
    // Update settings function
    let update_settings = Callback::new(move |update: AppSettings| {
        let settings_clone = update.clone();
        settings.set(update);
        
        // Persist to backend asynchronously
        spawn_local(async move {
            if let Err(e) = update_app_settings(&settings_clone).await {
                web_sys::console::warn_1(&format!("[SettingsPersistence]: failed to persist settings: {}", e).into());
            }
        });
    });
    
    (settings, is_hydrated, update_settings)
}

File: src/hooks/settings/usePersistentSettings.ts
import { useCallback, useEffect, useState } from "react";
import { invoke } from "@tauri-apps/api/core";
import { DEFAULT_KOKORO_VOICE_ID } from "../../constants/kokoro";
import type { AppSettings } from "../../types/settings";
import { logger } from "../../lib/logger";

const DEFAULT_SETTINGS: AppSettings = {
  theme: "system",
  ttsVoiceId: DEFAULT_KOKORO_VOICE_ID,
  autoScrollEnabled: true,
  audioPlaybackSpeed: 1.0,
};

/**
 * Hook for persisting app settings to backend database
 */
export function usePersistentSettings() {
  const [settings, setSettings] = useState<AppSettings>(DEFAULT_SETTINGS);
  const [isHydrated, setIsHydrated] = useState(false);

  // Load settings from backend on mount
  useEffect(() => {
    let cancelled = false;

    const loadSettings = async () => {
      // Add timeout to ensure hydration completes even if Tauri command hangs
      const timeoutId = setTimeout(() => {
        if (!cancelled) {
          logger.warn("[SettingsPersistence]: Settings load timeout, using defaults");
          setIsHydrated(true);
        }
      }, 5000); // 5 second timeout

      try {
        const loadedSettings = await invoke<AppSettings>("get_app_settings");
        clearTimeout(timeoutId);
        if (!cancelled) {
          // Merge with defaults to ensure all fields are present
          setSettings({ ...DEFAULT_SETTINGS, ...loadedSettings });
          setIsHydrated(true);
        }
      } catch (error) {
        clearTimeout(timeoutId);
        logger.warn("[SettingsPersistence]: failed to load settings from backend, using defaults:", { error });
        if (!cancelled) {
          setIsHydrated(true);
        }
      }
    };

    void loadSettings();

    return () => {
      cancelled = true;
    };
  }, []);

  // Update settings function
  const updateSettings = useCallback((update: Partial<AppSettings>) => {
    setSettings((prev) => {
      const newSettings = { ...prev, ...update };
      
      // Persist to backend asynchronously
      invoke("update_app_settings", { settings: newSettings }).catch((error) => {
        logger.warn("[SettingsPersistence]: failed to persist settings:", { error });
      });
      
      return newSettings;
    });
  }, []);

  return {
    settings,
    updateSettings,
    isHydrated,
  };
}


